### PR TITLE
feat: Add `--oauth-force-oob` CLI option

### PR DESCRIPTION
**Pull request by laurentsimon**
Adresses comment https://github.com/sigstore/sigstore-python/issues/665#issuecomment-1574033667

### CODE REVIEW

File: sigstore/_utils.py
Comment (laurentsimon) [inline sigstore/_utils.py:181]: this was failing the `make lint`, so I removed it

Comment (woodruffw) [reply sigstore/_utils.py:181]: > this was failing the make lint, so I removed it

Sorry, that's my fault -- the CI needs it, because we lint in CI against our minimum supported Python. So please add that back in 🙂

File: sigstore/_cli.py
Comment (woodruffw) [inline sigstore/_cli.py:225]: I think this flag name might be a little ambiguous: what it's doing is forcing the OOB flow, which is conceptually untied to the "default" browser. It's also doing an OAuth2 flow, so I think we should emphasize that rather than OIDC (OIDC refers to the "shape" of the claims that come out of the flow).

What do you think about `--oauth-no-browser` or `--oauth-force-oob`?

Comment (woodruffw) [reply sigstore/_cli.py:225]: (I have a slight preference for the latter, since it'll be consistent with the environment suggestion below. But either is fine IMO; your call.)

File: sigstore/oidc.py
Comment (woodruffw) [inline sigstore/oidc.py:255]: Let's update the docstring here as well, to emphasize that `force_oob` modifies the browser-opening behavior.

**Review by woodruffw — COMMENTED** [review]
Thanks for adding this! A couple of nitpicks 🙂

**Comment by woodruffw** [comment]
JFYI: This will also require a `CHANGELOG.md` entry and updates to the `--help` outputs in the `README.md`. But we can do that as the last step.

File: sigstore/_cli.py
Comment (laurentsimon) [reply sigstore/_cli.py]: I renamed `--oauth-force-oob`, even though I think prepending `--oidc-xxx` wold be more consistent with other commands if it's under the oidc group. Lmk

File: sigstore/oidc.py
Comment (laurentsimon) [reply sigstore/oidc.py:255]: PTAL

File: sigstore/_cli.py
Comment (woodruffw) [reply sigstore/_cli.py]: Yeah, it's a little inconsistent, but IMO it's worth it for the compatibility/symmetry with the existing flag. 

There might be a better section to put this under; I'll take a look.

File: sigstore/oidc.py
Comment (woodruffw) [reply sigstore/oidc.py:255]: Looks good, thank you!

**Comment by woodruffw** [comment]
LGTM behaviorally!

`check-readme` is unhappy, but I'll fix that:

```
# sigstore --help
# sigstore sign --help
4,5c4,5
<                      [--no-default-files] [--signature FILE]
make: *** [Makefile:[12](https://github.com/sigstore/sigstore-python/actions/runs/5158329157/jobs/9292054589?pr=667#step:5:13)3: check-readme] Error 1
<                      [--certificate FILE] [--bundle FILE]
---
>                      [--oauth-force-oob] [--no-default-files]
>                      [--signature FILE] [--certificate FILE] [--bundle FILE]
30c30,32
<   --oauth-force-oob     Force an out-of-band OAuth flow and do not automatically start the default web browser (default: False)
---
>   --oauth-force-oob     Force an out-of-band OAuth flow and do not
>                         automatically start the default web browser (default:
>                         False)
```

**Review by woodruffw — APPROVED** [review]

**Review by woodruffw — APPROVED** [review]
LGTM, thanks @laurentsimon! 

I need one of the other maintainers to approve as well, but this will be included in the 2.0 series.

**Review by di — APPROVED** [review]

**Review by di — DISMISSED** [review]

**Review by tetsuo-cpp — DISMISSED** [review]

**Comment by woodruffw** [comment]
I have no clue why the macOS and Windows CIs are getting stuck here. Maybe some kind of bad runner state? Nothing in this PR itself should vary between platforms.

**Comment by woodruffw** [comment]
Weirdly, both seem to hang at around the same test:

Windows:

```
2023-06-08T20:15:13.5528997Z test/unit/internal/oidc/test_issuer.py::test_fail_init_url PASSED        [ 75%]
2023-06-08T20:15:13.6085787Z test/unit/internal/oidc/test_issuer.py::test_init_url PASSED             [ 75%]
2023-06-08T20:22:23.5095829Z ##[error]The operation was canceled.
2023-06-08T20:22:23.6691779Z Post job cleanup.
2023-06-08T20:22:24.0268157Z [command]"C:\Program Files\Git\bin\git.exe" version
```

macOS:

```
2023-06-08T20:15:08.1732720Z test/unit/internal/oidc/test_issuer.py::test_fail_init_url PASSED        [ 75%]
2023-06-08T20:15:08.1998280Z test/unit/internal/oidc/test_issuer.py::test_init_url PASSED             [ 75%]
2023-06-08T20:22:31.5320710Z ##[error]The operation was canceled.
2023-06-08T20:22:31.5554660Z Post job cleanup.
2023-06-08T20:22:32.1636730Z [command]/usr/local/bin/git version
2023-06-08T20:22:32.2323580Z git version 2.40.1
```

I'll do some more digging later.

**Comment by di** [comment]
I wonder if these are waiting for user input... perhaps we need to add a reasonable timeout around any flows that wait for user input?

**Comment by woodruffw** [comment]
> I wonder if these are waiting for user input... perhaps we need to add a reasonable timeout around any flows that wait for user input?

Yeah, that was my guess, although it's a mystery to me why they'd block on these platforms and not on the Linux CI. 

I think a timeout around user input makes sense, in places we can enforce it (e.g. `stdin` and not the browser) -- I'll look into that as well.

**Comment by woodruffw** [comment]
This was indeed a silly blocking thing; should be fixed with d6052c9.

**Comment by di** [comment]
How was this passing on linux? E.g. https://github.com/sigstore/sigstore-python/actions/runs/5204665487/jobs/9412771815

**Comment by woodruffw** [comment]
I think it probably passed because we're not being specific enough about _which_ `IdentityError` is being caught in the test.

Looking at it more closely, I'm not 100% sure what branch this test was originally intended to cover -- @jleightcap do you remember what your intent with this was?

**Comment by woodruffw** [comment]
It looks like it was probably meant to catch this:

https://github.com/sigstore/sigstore-python/blob/d61025510f6497bf82fce89cd2aec27383c64cfe/sigstore/oidc.py#L361-L364

I'll add a proper exception message there and assert on it. My guess is that the Linux CI was passing because `webbrowser.open()` failed on it and pushed it into the OOB flow, even without `force_oob`.

**Comment by woodruffw** [comment]
Tests are passing with the new match, so that confirms that the Linux CI was coincidentally entering the OOB flow (whereas Windows and macOS weren't, probably because both have a browser installed).

**Review by di — APPROVED** [review]

**Pull request merged**
